packages/core/src/entities/feature-descriptor.ts

```diff
@@ -0,0 +1,143 @@
+import { SetMultiMap } from '@wixc3/patterns';
+import type { RuntimeEngine } from '../runtime-engine';
+import type { ENGINE, RUN_OPTIONS } from '../symbols';
+import type {
+    Context,
+    DeepEnvironmentDeps,
+    EntityRecord,
+    EnvVisibility,
+    FilterNotEnv,
+    GetDependenciesOutput,
+    GetOnlyLocalUniversalOutputs,
+    GetRemoteOutputs,
+    IRunOptions,
+    MapRecordType,
+    MapToProxyType,
+    MapType,
+    MapVisibleInputs,
+    OmitCompositeEnvironment,
+    PartialFeatureConfig,
+    RegisteringFeature,
+    Running,
+} from '../types';
+import type { AnyEnvironment, GloballyProvidingEnvironments } from './env';
+
+export function createRuntimeInfo(): RuntimeInfo {
+    return {
+        setup: new SetMultiMap(),
+        context: new Map(),
+        envs: new Set(),
+    };
+}
+
+export function provideConfig<T extends FeatureDescriptor>(
+    feature: T,
+    config: PartialFeatureConfig<T['api']>
+): [T['id'], PartialFeatureConfig<T['api']>] {
+    return [feature.id, config];
+}
+
+export function setup<T extends FeatureDescriptor, E extends AnyEnvironment>(
+    feature: T,
+    environment: E,
+    setupHandler: SetupHandler<T, E>
+) {
+    const info = (feature.runtimeInfo ||= createRuntimeInfo());
+    validateNoDuplicateEnvRegistration(environment, feature.id, info.envs);
+    info.setup.add(environment.env, setupHandler);
+}
+
+export function setupContext<T extends FeatureDescriptor, E extends AnyEnvironment, K extends keyof T['context']>(
+    feature: T,
+    _environment: E, // TODO: add handlers in environments buckets with validation per environment?
+    environmentContextKey: K,
+    contextHandler: ContextHandler<T, E, K>
+) {
+    const info = (feature.runtimeInfo ||= createRuntimeInfo());
+    validateNoDuplicateContextRegistration(environmentContextKey, feature.id, info.context);
+    info.context.set(environmentContextKey, contextHandler);
+}
+
+export function validateNoDuplicateEnvRegistration(env: AnyEnvironment, featureId: string, registered: Set<string>) {
+    const hasCollision = testEnvironmentCollision(env, registered);
+    if (hasCollision.length) {
+        const collisions = hasCollision.join(', ');
+        throw new Error(
+            `Feature can only have single setup for each environment. ${featureId} Feature already implements: ${collisions}`
+        );
+    }
+}
+
+export function testEnvironmentCollision(envVisibility: EnvVisibility, envSet: Set<string>): string[] {
+    const containsEnv = new Set<string>();
+    const test = (env: string) => {
+        envSet.has(env) ? containsEnv.add(env) : envSet.add(env);
+    };
+    if (Array.isArray(envVisibility)) {
+        for (const e of envVisibility) {
+            test(e.env);
+        }
+    } else if (typeof envVisibility === 'string') {
+        test(envVisibility);
+    } else {
+        test(envVisibility.env);
+    }
+    return [...containsEnv];
+}
+
+export function validateNoDuplicateContextRegistration(
```

Comment:
inline the contents of this to the only place using it (setupContext)

Reply:
This currently used in two places when I delete the old feature I will inline. might be in this PR (but later stage).

Reply:
still used in multiple places... when we delete Feature I will consider inline it